Lib/traceback.py

```diff
@@ -486,56 +504,138 @@ def format_frame_summary(self, frame_summary):
-                        row.append(anchors.secondary_char * (dp_right_start_offset - dp_left_end_offset))
-                        row.append(anchors.primary_char * (dp_end_offset - dp_start_offset - dp_right_start_offset))
-                    else:
-                        row.append('^' * (dp_end_offset - dp_start_offset))
+                # only output first line if column information is missing
+                row.append(textwrap.indent(frame_summary.line, '    ') + "\n")
+            else:
+                # get first and last line
+                all_lines_original = frame_summary._original_line.splitlines()
+                first_line = all_lines_original[0]
+                last_line = all_lines_original[frame_summary.end_lineno - frame_summary.lineno]
+
+                # character index of the start/end of the instruction
+                start_offset = _byte_offset_to_character_offset(first_line, frame_summary.colno)
+                end_offset = _byte_offset_to_character_offset(last_line, frame_summary.end_colno)
+
+                all_lines = frame_summary._dedented_lines.splitlines()[
+                    :frame_summary.end_lineno - frame_summary.lineno + 1
+                ]
 
-                    row.append('\n')
+                # adjust start/end offset based on dedent
+                dedent_characters = len(first_line) - len(all_lines[0])
+                start_offset -= dedent_characters
+                end_offset -= dedent_characters
+                start_offset = max(0, start_offset)
+                end_offset = max(0, end_offset)
+
+                # When showing this on a terminal, some of the non-ASCII characters
+                # might be rendered as double-width characters, so we need to take
+                # that into account when calculating the length of the line.
+                dp_start_offset = _display_width(all_lines[0], offset=start_offset)
+                dp_end_offset = _display_width(all_lines[-1], offset=end_offset)
+
+                # get exact code segment corresponding to the instruction
+                segment = "\n".join(all_lines)
+                segment = segment[start_offset:len(segment) - (len(all_lines[-1]) - end_offset)]
+
+                # attempt to parse for anchors
+                anchors: Optional[_Anchors] = None
+                try:
+                    anchors = _extract_caret_anchors_from_line_segment(segment)
+                except Exception:
+                    pass
+
+                # only use carets if there are anchors or the carets do not span all lines
+                show_carets = False
+                if anchors or all_lines[0][:start_offset].lstrip() or all_lines[-1][end_offset:].rstrip():
+                    show_carets = True
+
+                result = []
+
+                # only display first line, last line, and lines around anchor start/end
+                significant_lines = {0, len(all_lines) - 1}
```

Comment:
I am not very convinced about this `significant_lines` treatment. I have the following concerns:

* It increases the complexity of the traceback
* There are plenty of cases where this hides seemingly arbitrary lines. Consider this code:
```
def foo(a,b,c,d,e,f,g):
    bar(b)

def bar(b):
    1/b

foo(
        1,
        0,
        3,
        4,
        5,
        6,
        7)
```

Currently this shows:

```
Traceback (most recent call last):
  File "/Users/pgalindo3/github/python/main/lel.py", line 7, in <module>
    foo(
    ~~~^
            1,
            ^^
    ...<4 lines>...
            6,
            ^^
            7)
            ^^
  File "/Users/pgalindo3/github/python/main/lel.py", line 2, in foo
    bar(b)
    ~~~^^^
  File "/Users/pgalindo3/github/python/main/lel.py", line 5, in bar
    1/b
    ~^~
ZeroDivisionError: division by zero
```

but it's not clear to me why we are hiding 4 lines there. Why are the arguments provided there not important? If you look closely what's causing the exception is that b is 0, but that's not being shown in the call to `foo`, but is the most important part of the call.

Reply:
That's true, although I agree with the general direction of the code to hide some lines, in case the error spans many lines and the traceback becomes unreadable. Should we maybe think of another heuristic to figure out which lines to show and which not to show?

Reply:
We will need some kind of heuristic - I encountered a case where a traceback through a `with` statement results in the entire block being printed because the instruction being traced points to the entire block.

Reply:
Do you have an example of the `with` statement out of curiosity?

Reply:
Code:
```python
class C:
    def __enter__(self):
        raise RuntimeError("fail")

    def __exit__(self, *excinfo):
        pass

with C():
    # comment
    # comment
    # comment
    # comment
    # comment
    # comment
    # comment
    # comment
    # comment
    pass
```
Output:
```
Traceback (most recent call last):
  File "/data/users/williamwen/cpython/local_test.py", line 8, in <module>
    with C():
    ...<9 lines>...
        pass
  File "/data/users/williamwen/cpython/local_test.py", line 3, in __enter__
    raise RuntimeError("fail")
RuntimeError: fail
```
So if we don't skip lines, we would print out all those comments.

Reply:
Agreed, let's go with this and we can refine it in the future if the need arises.

Reply:
Indeed, looks like all the bytecode instruction for the with include all the comments:

```
[]
Instruction(opname='RESUME', opcode=151, arg=0, argval=0, argrepr='', offset=0, starts_line=0, is_jump_target=False, positions=Positions(lineno=0, end_lineno=1, col_offset=0, end_col_offset=0))
['class C:', '    def __enter__(self):', '        raise RuntimeError("fail")', '', '    def __exit__(self, *excinfo):']
Instruction(opname='PUSH_NULL', opcode=2, arg=None, argval=None, argrepr='', offset=2, starts_line=1, is_jump_target=False, positions=Positions(lineno=1, end_lineno=6, col_offset=0, end_col_offset=12))
['class C:', '    def __enter__(self):', '        raise RuntimeError("fail")', '', '    def __exit__(self, *excinfo):']
Instruction(opname='LOAD_BUILD_CLASS', opcode=71, arg=None, argval=None, argrepr='', offset=4, starts_line=None, is_jump_target=False, positions=Positions(lineno=1, end_lineno=6, col_offset=0, end_col_offset=12))
['class C:', '    def __enter__(self):', '        raise RuntimeError("fail")', '', '    def __exit__(self, *excinfo):']
Instruction(opname='LOAD_CONST', opcode=100, arg=0, argval=<code object C at 0x1049211b0, file "<string>", line 1>, argrepr='<code object C at 0x1049211b0, file "<string>", line 1>', offset=6, starts_line=None, is_jump_target=False, positions=Positions(lineno=1, end_lineno=6, col_offset=0, end_col_offset=12))
['class C:', '    def __enter__(self):', '        raise RuntimeError("fail")', '', '    def __exit__(self, *excinfo):']
Instruction(opname='MAKE_FUNCTION', opcode=132, arg=0, argval=0, argrepr='', offset=8, starts_line=None, is_jump_target=False, positions=Positions(lineno=1, end_lineno=6, col_offset=0, end_col_offset=12))
['class C:', '    def __enter__(self):', '        raise RuntimeError("fail")', '', '    def __exit__(self, *excinfo):']
Instruction(opname='LOAD_CONST', opcode=100, arg=1, argval='C', argrepr="'C'", offset=10, starts_line=None, is_jump_target=False, positions=Positions(lineno=1, end_lineno=6, col_offset=0, end_col_offset=12))
['class C:', '    def __enter__(self):', '        raise RuntimeError("fail")', '', '    def __exit__(self, *excinfo):']
Instruction(opname='PRECALL', opcode=166, arg=2, argval=2, argrepr='', offset=12, starts_line=None, is_jump_target=False, positions=Positions(lineno=1, end_lineno=6, col_offset=0, end_col_offset=12))
['class C:', '    def __enter__(self):', '        raise RuntimeError("fail")', '', '    def __exit__(self, *excinfo):']
Instruction(opname='CALL', opcode=171, arg=2, argval=2, argrepr='', offset=16, starts_line=None, is_jump_target=False, positions=Positions(lineno=1, end_lineno=6, col_offset=0, end_col_offset=12))
['class C:', '    def __enter__(self):', '        raise RuntimeError("fail")', '', '    def __exit__(self, *excinfo):']
Instruction(opname='STORE_NAME', opcode=90, arg=0, argval='C', argrepr='C', offset=26, starts_line=None, is_jump_target=False, positions=Positions(lineno=1, end_lineno=6, col_offset=0, end_col_offset=12))
[]
Instruction(opname='PUSH_NULL', opcode=2, arg=None, argval=None, argrepr='', offset=28, starts_line=8, is_jump_target=False, positions=Positions(lineno=8, end_lineno=8, col_offset=5, end_col_offset=6))
[]
Instruction(opname='LOAD_NAME', opcode=101, arg=0, argval='C', argrepr='C', offset=30, starts_line=None, is_jump_target=False, positions=Positions(lineno=8, end_lineno=8, col_offset=5, end_col_offset=6))
[]
Instruction(opname='PRECALL', opcode=166, arg=0, argval=0, argrepr='', offset=32, starts_line=None, is_jump_target=False, positions=Positions(lineno=8, end_lineno=8, col_offset=5, end_col_offset=8))
[]
Instruction(opname='CALL', opcode=171, arg=0, argval=0, argrepr='', offset=36, starts_line=None, is_jump_target=False, positions=Positions(lineno=8, end_lineno=8, col_offset=5, end_col_offset=8))
['with C():', '    # comment', '    # comment', '    # comment', '    # comment', '    # comment', '    # comment', '    # comment', '    # comment', '    # comment']
Instruction(opname='BEFORE_WITH', opcode=53, arg=None, argval=None, argrepr='', offset=46, starts_line=None, is_jump_target=False, positions=Positions(lineno=8, end_lineno=18, col_offset=0, end_col_offset=8))
['with C():', '    # comment', '    # comment', '    # comment', '    # comment', '    # comment', '    # comment', '    # comment', '    # comment', '    # comment']
Instruction(opname='POP_TOP', opcode=1, arg=None, argval=None, argrepr='', offset=48, starts_line=None, is_jump_target=False, positions=Positions(lineno=8, end_lineno=18, col_offset=0, end_col_offset=8))
[]
Instruction(opname='NOP', opcode=9, arg=None, argval=None, argrepr='', offset=50, starts_line=18, is_jump_target=False, positions=Positions(lineno=18, end_lineno=18, col_offset=4, end_col_offset=8))
['with C():', '    # comment', '    # comment', '    # comment', '    # comment', '    # comment', '    # comment', '    # comment', '    # comment', '    # comment']
Instruction(opname='LOAD_CONST', opcode=100, arg=2, argval=None, argrepr='None', offset=52, starts_line=8, is_jump_target=False, positions=Positions(lineno=8, end_lineno=18, col_offset=0, end_col_offset=8))
['with C():', '    # comment', '    # comment', '    # comment', '    # comment', '    # comment', '    # comment', '    # comment', '    # comment', '    # comment']
Instruction(opname='LOAD_CONST', opcode=100, arg=2, argval=None, argrepr='None', offset=54, starts_line=None, is_jump_target=False, positions=Positions(lineno=8, end_lineno=18, col_offset=0, end_col_offset=8))
['with C():', '    # comment', '    # comment', '    # comment', '    # comment', '    # comment', '    # comment', '    # comment', '    # comment', '    # comment']
Instruction(opname='LOAD_CONST', opcode=100, arg=2, argval=None, argrepr='None', offset=56, starts_line=None, is_jump_target=False, positions=Positions(lineno=8, end_lineno=18, col_offset=0, end_col_offset=8))
['with C():', '    # comment', '    # comment', '    # comment', '    # comment', '    # comment', '    # comment', '    # comment', '    # comment', '    # comment']
Instruction(opname='PRECALL', opcode=166, arg=2, argval=2, argrepr='', offset=58, starts_line=None, is_jump_target=False, positions=Positions(lineno=8, end_lineno=18, col_offset=0, end_col_offset=8))
['with C():', '    # comment', '    # comment', '    # comment', '    # comment', '    # comment', '    # comment', '    # comment', '    # comment', '    # comment']
Instruction(opname='CALL', opcode=171, arg=2, argval=2, argrepr='', offset=62, starts_line=None, is_jump_target=False, positions=Positions(lineno=8, end_lineno=18, col_offset=0, end_col_offset=8))
['with C():', '    # comment', '    # comment', '    # comment', '    # comment', '    # comment', '    # comment', '    # comment', '    # comment', '    # comment']
Instruction(opname='POP_TOP', opcode=1, arg=None, argval=None, argrepr='', offset=72, starts_line=None, is_jump_target=False, positions=Positions(lineno=8, end_lineno=18, col_offset=0, end_col_offset=8))
['with C():', '    # comment', '    # comment', '    # comment', '    # comment', '    # comment', '    # comment', '    # comment', '    # comment', '    # comment']
Instruction(opname='LOAD_CONST', opcode=100, arg=2, argval=None, argrepr='None', offset=74, starts_line=None, is_jump_target=False, positions=Positions(lineno=8, end_lineno=18, col_offset=0, end_col_offset=8))
['with C():', '    # comment', '    # comment', '    # comment', '    # comment', '    # comment', '    # comment', '    # comment', '    # comment', '    # comment']
Instruction(opname='RETURN_VALUE', opcode=83, arg=None, argval=None, argrepr='', offset=76, starts_line=None, is_jump_target=False, positions=Positions(lineno=8, end_lineno=18, col_offset=0, end_col_offset=8))
['with C():', '    # comment', '    # comment', '    # comment', '    # comment', '    # comment', '    # comment', '    # comment', '    # comment', '    # comment']
Instruction(opname='PUSH_EXC_INFO', opcode=35, arg=None, argval=None, argrepr='', offset=78, starts_line=None, is_jump_target=False, positions=Positions(lineno=8, end_lineno=18, col_offset=0, end_col_offset=8))
['with C():', '    # comment', '    # comment', '    # comment', '    # comment', '    # comment', '    # comment', '    # comment', '    # comment', '    # comment']
Instruction(opname='WITH_EXCEPT_START', opcode=49, arg=None, argval=None, argrepr='', offset=80, starts_line=None, is_jump_target=False, positions=Positions(lineno=8, end_lineno=18, col_offset=0, end_col_offset=8))
['with C():', '    # comment', '    # comment', '    # comment', '    # comment', '    # comment', '    # comment', '    # comment', '    # comment', '    # comment']
Instruction(opname='POP_JUMP_FORWARD_IF_TRUE', opcode=115, arg=4, argval=92, argrepr='to 92', offset=82, starts_line=None, is_jump_target=False, positions=Positions(lineno=8, end_lineno=18, col_offset=0, end_col_offset=8))
['with C():', '    # comment', '    # comment', '    # comment', '    # comment', '    # comment', '    # comment', '    # comment', '    # comment', '    # comment']
Instruction(opname='RERAISE', opcode=119, arg=2, argval=2, argrepr='', offset=84, starts_line=None, is_jump_target=False, positions=Positions(lineno=8, end_lineno=18, col_offset=0, end_col_offset=8))
```

@iritkatriel I think this qualifies as a bug. Fo example the `CALL` for `C()` in `with C()` has `end_lineno` setted to the last comment in the block.